Wip #19442 add NotUsed and Done types

johanandren opened the pull request:
Refs #19442

rkuhn commented on akka-actor/src/main/scala/akka/NotUsed.scala:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Copyright (C) 2009-2016 Typesafe Inc. <http://www.typesafe.com>
```
I thought this should just mention 2016?

johanandren replied:
You are probably right. Was cargo cult copyrighting ;)

rkuhn commented on akka-actor/src/main/scala/akka/NotUsed.scala:
```diff
@@ -0,0 +1,20 @@
+
+/**
+ * A value is not intended to be used.
+ */
```
This comment should be removed: it is enough to document the type.

drewhk replied:
It doesn't hurt to repeat it though.

rkuhn replied:
Then it should be a copy of the other one, though.

rkuhn commented on akka-actor/src/main/scala/akka/NotUsed.scala:
````diff
@@ -0,0 +1,20 @@
+
+/**
+ * Type of value is not intended to be used. Is more clear than Unit and available
+ * both from Scala and Java (which ```Unit``` is not).
````
“This type is used in generic type signatures wherever the actual value is of no importance. It is a combination of Scala’s Unit and Java’s Void, which both have different issues when used from the other language. An example use-case is the materialized value of an Akka Stream for cases where no result shall be returned from materialization.”

ktoso commented on akka-actor/src/main/scala/akka/Done.scala:
```diff
@@ -0,0 +1,16 @@
+
+case object Done extends Done {
+  def getInstance() = this
+}
```
Technically Java people can now `new Done()`. 
Did we consider `enum`? What are the downsides?

rkuhn replied:
We could assert in the constructor that initialization happens from within the `object` context.

drewhk replied:
originally I created an enum, but that is not as nice in Scala since you won't be able to use `Done` (although `Done()` is possible by adding a static `apply()` on the Enum)

ktoso replied:
Ah right, so maybe the require Roland mentions?

rkuhn commented on akka-docs/rst/project/migration-guide-streams-2.0.x-2.4.x.rst:
```diff
@@ -6,3 +6,41 @@ Migration Guide Akka Streams 2.0.x to 2.4.x
+
+``akka.Done`` is added for the use case where it is boxed inside another object to signify
+completion but there is no actual value attached to the completion. It is used to replace
+occurrences of ``Future<BoxedUnit>`` with ``Future<Done>``in Java and ``Future[Unit]`` with
```
missing space after code

rkuhn commented on akka-docs/rst/project/migration-guide-streams-2.0.x-2.4.x.rst:
```diff
@@ -6,3 +6,41 @@ Migration Guide Akka Streams 2.0.x to 2.4.x
+And Scala code like this::
+
+    Source[Int, Unit] source = Source.from(1 to 5)
+    Sink[Int, Future[Unit]] sink = Sink.ignore()
```
… needs to be changed to? :-)

ktoso commented on akka-actor/src/main/scala/akka/Done.scala:
````diff
@@ -0,0 +1,16 @@
+/**
+ * Typically used together with ```Future``` to signal completion
+ * but there is no actual value completed. More clearly signals intent
+ * than ```Unit``` and is available both from Scala and Java (which ```Unit``` is not).
````
trilpe `? Another feature of ScalaDoc I'm unaware of? Thought the `usual` works hmm

johanandren replied:
Oh, I only knew of the triple one. (Which is super annoying on a swedish keyboard because it wants to combine with the previous character)

drewhk commented on akka-docs/rst/project/migration-guide-streams-2.0.x-2.4.x.rst:
```diff
@@ -6,3 +6,41 @@ Migration Guide Akka Streams 2.0.x to 2.4.x
+
+    Source[Int, Unit] source = Source.from(1 to 5)
+    Sink[Int, Future[Unit]] sink = Sink.ignore()
+
```
"needs to be changed into"

drewhk commented:
LGTM

rkuhn commented:
LGTM

patriknw commented:
LGTM, very nice!

drewhk commented:
sorry, it didn't compile :(

drewhk commented:
still not :(

drewhk commented:
and no longer merges either

johanandren commented:
`clean` `compile` `test:compile` before I push again!

rkuhn commented:
do you want to fix it up or shall I? I can do the “I’m the boss and push directly to master” thing :-) (well, I’d have to “override all the safety checks” as they say in the movies)

rkuhn commented:
I might have broken compilation by clicking a green button a few minutes ago … sorry.

drewhk commented:
no, because it started validation way before your merge

drewhk commented:
If you can, please push this over the finish line, as this will conflict with everything otherwise. Thankfully the PR queue is now containing only "waiting for action" items, so no further merges are expeced.

johanandren commented:
Will try, or let you know.

johanandren commented:
Nevereeending story

2m commented on akka-docs/rst/project/migration-guide-streams-2.0.x-2.4.x.rst:
```diff
@@ -6,3 +6,47 @@ Migration Guide Akka Streams 2.0.x to 2.4.x
+
+akka.Done and akka.NotUsed replacing Unit and BoxedUnit
+-------------------------------------------------------
+To provide a more clear signatures and have a unified API for both
```
s/a more clear/more clear

2m commented on akka-docs/rst/project/migration-guide-streams-2.0.x-2.4.x.rst:
```diff
@@ -6,3 +6,47 @@ Migration Guide Akka Streams 2.0.x to 2.4.x
+akka.Done and akka.NotUsed replacing Unit and BoxedUnit
+-------------------------------------------------------
+To provide a more clear signatures and have a unified API for both
+Java and Scala two new types has been introduced:
```
s/has/have

2m commented on akka-docs/rst/project/migration-guide-streams-2.0.x-2.4.x.rst:
```diff
@@ -6,3 +6,47 @@ Migration Guide Akka Streams 2.0.x to 2.4.x
+``akka.NotUsed`` is meant to be used instead of ``Unit`` in Scala
+and ``BoxedUnit`` in Java to signify that the type parameter is required
+but not actually used. This is commonly the case with ``Source``s, ``Flow``s and ``Sink``s
+that does not materialize into any value.
```
s/does/do

2m commented on akka-docs/rst/project/migration-guide-streams-2.0.x-2.4.x.rst:
```diff
@@ -6,3 +6,47 @@ Migration Guide Akka Streams 2.0.x to 2.4.x
+    Source[Int, NotUsed] source = Source.from(1 to 5)
+    Sink[Int, Future[Done]] sink = Sink.ignore()
+
+These changes applies to all the places where streams are used, which means that signatures
```
s/applies/apply

2m commented:
Rebased and fixed compilations errors as this is needed for the work on #19446.

I did not want to take over the committer credits, but `git rebase master` did that automatically, so...

patriknw commented:
you can change author with `git commit --amend --author=...`

2m commented:
Great. When/if the validation passes I will push the right commit and merge instantly without allowing validation to kick in again.

Commit:
also fix FlowInterleaveSpec

Also-by: Johan Andrén <johan@markatta.com>
Also-by: Roland Kuhn <rk@rkuhn.info>
Also-by: Martynas Mickevičius <mmartynas@gmail.com>
